Added first test version of parser.

sebasguts opened the pull request.

kalmarek commented on deps/parser/parser.jl:
```diff
@@ -0,0 +1,184 @@
+using JSON
+
+typestring = ""
+undotted_typestring = typestring[1:length(typestring)-1]
```
what is typestring intended for? future namespacing?
If You need to get rid of the dot at the end use `strip(typestring, ".")`

sebasguts replied:
Yep. We have not yet resolved the internal module. It was previously used in different places, I guess we can actually remove it now.

kalmarek commented on deps/parser/parser.jl:
```diff
@@ -0,0 +1,184 @@
+    ## We only translate the known cases
+    if !haskey(type_translate_list,type_string)
+        println("Unknown type: " * type_string)
+        return false
```
here we could just throw an 
```julia
struct UnknownPolymakeType <: Exception
    msg::String
    UnknownPolymakeType(type_string) = new("Encountered an unknown polymake type: $type_string")
end
```

sebasguts replied:
This is not supposed to throw an exception, this is just supposed to warn me.

kalmarek replied:
then `@warn(...)`;
but You check for `falses` above, so this may as well throw with handling doing the warning

kalmarek commented on deps/parser/parser.jl:
```diff
@@ -0,0 +1,184 @@
+    argument_list = Array{String,1}()
+
+    for i in 1:max_argument_number
+        push!(argument_list_header,create_argument_string_from_type(arguments[i],i))
```
```julia
try
    push!(argument_list_header,create_argument_string_from_type(arguments[i],i))
catch ex
    if ex isa UnknownPolymakeType
         ...
    else
         rethrow(ex)
    end
end

sebasguts replied:
See above

kalmarek commented on deps/parser/parser.jl:
```diff
@@ -0,0 +1,184 @@
+        end
+    end
+
+    return_string = ""
```
this is probably the place to separate json parsing from code generation

sebasguts replied:
Please elaborate?

kalmarek replied:
below we mostly fill the return_string with all the data we pulled (and transformed) from json;
and the name of the function reads `parse...`, whereas it does both parsing and code gen

sebasguts replied:
K.

sebasguts replied:
I decided to not do this at the moment. The name of the function in the current state is probably debatable, but currently there is really no use in ripping this function apart into two.

Granted, two would be a bit more clean, but you end up in having to parse all the extracted information from the first part to the second part. Either using keyword arguments or a dict, since everything else would be very messy.

But since both parts are resonably small, I would leave them as they are.

kalmarek commented on deps/parser/parser.jl:
```diff
@@ -0,0 +1,184 @@
+"""
+    for current_function in parsed_dict["functions"]
+        return_value = parse_function_definition(current_function,app_name)
+        if return_value == false
```
or even catch that exception again here

kalmarek commented on deps/parser/parser.jl:
```diff
@@ -0,0 +1,184 @@
+            if argument_list_header[i] == false
+                if i <= min_argument_number
+                    println("Cannot parse function " * name * " due to unparsable argument")
+                    return false
```
a pretty terrible way of signaling a failure ;-)

sebasguts replied:
As said above, I wanted to have this information printed, not really providing any exception handling. If Julia has a standard way to print warnings, we could use that one instead. I definitely do not want an error here.

kalmarek replied:
if You catch you can just print a warning and continue as if nothing has happened

sebasguts commented on deps/parser/parser.jl:
```diff
@@ -0,0 +1,184 @@
+        end
+    end
+    return_string = return_string * "\n\nend\nexport $app_name\n"
+    open(abspath(outputfileposition * "/" * app_name * ".jl" ),"w") do outputfile
```
Does `joinpath` also work on windows?

kalmarek commented:
Did You think about separating the flow of `parse_function_definition` for methods and functions?

I'd probably made a types `PolymakeFunction`/ `PolymakeMethod` and dispatch code generation on them

but we can work on this later, if You want to have this merged fast

sebasguts commented:
I thought about separating the flow first, but the changes were too minimal. I wanted to avoid unnecessary code duplication.

kalmarek commented:
@sebasguts could You also push some json with methods? the one in this pull has only functions

saschatimme commented on deps/parser/parser.jl:
```diff
@@ -0,0 +1,184 @@
+    return """
+function $julia_name( $julia_args )
+    $(typestring)application( "$app_name" )
+    return $(typestring)call_method( "$polymake_name$parameter_string", dispatch_obj, Array{Any,1}([ $call_args ]) )
```
Is this `dispatch_obj` really always the same? Will this implicitly be in the `julia_args` list?

saschatimme replied:
I see this is implictly provided

sebasguts replied:
Yep, it is always the same. The methods are object methods.

saschatimme commented:
For me it is also fine if we merge the parser for now (such that Sebastian can do other stuff) and I prettify everything afterwards. Since Sebastian probably has more important stuff to do than producing pretty Julia code. That's something where I can help out.

kalmarek commented:
I am fine with this as well, @sebasguts just push more examples.

now, where do I approve this thing?

sebasguts commented:
I will do the example stuff and some minor refactoring now, just because I also need to learn how to produce pretty Julia code. Full blown exception handling is a bit overkill here, but I probably have just worked with too many languages that do not have any, so I need to get used to it again. I will push my changes tonight, then you can have another look and we merge this stuff tomorrow.

I am not completely sure this will be the finished thing, I still need to wait for some more JSON to see if I found all corner cases (Spoiler alert: Probably not). Afterwards, once the JSON is there, we can mock up some test cases.

sebasguts commented:
So, I have added the exceptions, plus minor comments from @kalmarek (Using `strip` and `joinpath`, mostly). I also renamed the large function into `parse_definition`. Tomorrow I will make up more JSON to test more corner cases, and specially the method case. I would suggest to not merge this PR yet. You can comment and review, but I would suggest merging it once it is finished.

kalmarek commented:
following Saschas comment I'm for merging it and prettifying later

sebasguts commented:
It is not really about prettifying later, but about fixing it up with more examples and some more changes from the polymake side. As said, I would rather continue to work on this branch until it is really somehow "finished", at least beta or something.

kalmarek commented:
could You share a json with "method"?

kalmarek commented on deps/parser/parser.jl:
```diff
@@ -0,0 +1,203 @@
+    has_option_set = last(arguments) == "OptionSet"
+
+    ## Make type parameters
+    param_list_header = Array{String,1}()
```
this could be made into a separate function:
```julia
function params(n::Int)
    if n == 0
        return Array{String,1}(), ""
    end

    param_list_header = Array{String,1}()
    param_list = Array{String,1}()

    for i in 1:n
        push!(param_list_header,"param$i::String")
        push!(param_list,"\$param$i")
    end

    param_string = "<" * join(param_list,",") * ">"

    return param_list_header, param_string
end
```

kalmarek commented on deps/parser/parser.jl:
```diff
@@ -0,0 +1,203 @@
+    end
+
+    ## Compute the argument range
+    max_argument_number = length(arguments) - (has_option_set ? 1 : 0)
```
```julia
max_args(len::Int, has_option::Bool) = len - (has_option ? 1 : 0)
```
then
```
if is_method
    min_arg_number = mandatory + 1
    min_arg_number = max_args(length(args), has_option_set) + 1
    argument_list_header = Array{String,1}(["dispatch_obj::$(typestring)pm_perl_Object"])
else # is_function
    min_arg_number = mandatory
    min_arg_number = max_args(length(args), has_option_set)
    argument_list_header = Array{String,1}()
end
```

kalmarek commented on deps/parser/parser.jl:
```diff
@@ -0,0 +1,203 @@
+    end
+    argument_list = Array{String,1}()
+
+    for i in 1:max_argument_number
```
I'd be best to make it into a `function argument_list(arguments, max_argument_number)` which
`return argument_list_header, argument_list`

kalmarek commented on deps/parser/parser.jl:
```diff
@@ -0,0 +1,203 @@
+                rethrow(exception)
+            end
+        end
+        push!(argument_list,"arg$i")
```
this is of little importance here, but repeated pushing is not the best thing here; comprehension ala
`argument_list = Vector{String}(["arg$i" for i in 1:max_argument_number])`
is better

saschatimme replied:
You can even drop the `Vector{String}` constructor.

kalmarek commented:
@sebasguts also the julian way to handle this would be to create separate types:
```julia
abstract type PMCallable end

struct PMFunction <: PMCallable
    app::String
    name::String
    args::Vector{String}
    mandatory::Int
    type_params::Int
end

struct PMMethod <: PMCallable
    app::String
    name::String
    args::Vector{String}
    mandatory::Int
    type_params::Int
end

function PMCallable(app_name::String, method_dict::Dict)
    if haskey(callable_dict, "method_of")
        return PMMethod(app_name, callable_dict)
    else
        return PMFunction(app_name, callable_dict)
    end
end
```
and then instead of ifing dispatch function on the type:

```julia
julia_name(c::PMCallable) = c.name
polymake_name(c::PMCallable) = c.name

has_option_set(c::PMCallable) = last(c.args) == "OptionSet"

max_args(len::Int, has_option::Bool) = len - (has_option ? 1 : 0)

max_arg_number(f::PMFunction) = max_args(length(f.args), has_option_set(f))
min_arg_number(f::PMFunction) = c.mandatory

min_arg_number(m::PMMethod) = c.mandatory + 1
min_arg_number(m::PMMethod) = max_args(length(c.args), has_option_set(c)) + 1
```
And so on.
We (Sascha and I) are more used to writing such code, so we could PR that for You, if You wish

saschatimme commented on deps/parser/parser.jl:
```diff
@@ -30,26 +30,39 @@ type_translate_list = Dict(
+    $(typestring)application( "$app_name" )
+    return_value = $(typestring)call_function( "$polymake_name$parameter_string", Array{Any,1}([ $call_args ]) )
+    return convert_from_property_value(return_value)
+end
```
You could also have just one function definition, the compiler figures that out.

```julia
function $julia_name( $julia_args , ::Val{Convert}=Val(true))
     $(typestring)application( "$app_name" )
     return_value = $(typestring)call_function( "$polymake_name$parameter_string", Array{Any,1}([ $call_args ]) )
     if Convert
         return convert_from_property_value(return_value)
    else
        return return_value
    end
end
```

sebasguts replied:
I see this works. I will do this.

sebasguts commented:
Closing this for a new one
